### DISCUSSION TITLE
How to transform velocity vectors between different coordinates?

### DISCUSSION THREAD

**Discussion by tiger2017** [spacepy/spacepy#563] [Support and Q&A] [answered]
Hello,

I am wondering how to transform the vector field (e.g. flows in the ionosphere: Vx, Vy, Vz, where they are velocity of eastward, northward and vertical direction) from the GEO to MAG coordinate system. I know how to transform the locations (e.g  latitude, longitude) by using the Coords function, but not sure how to transform the vectors between different coordinate systems. Thank you.

**Comment by jtniehof** [comment]
Anything that's a simple rotation (e.g. magnetic field) can be treated by "pretending" it's a location. Velocities though are a different matter, since the full six-dimension state is needed. @drsteve , am I missing something? Should we set up an enhancement once #296 is merged?

**Comment by drsteve** [comment]
There are some subtleties to this one (remembering that position is a vector, just with the origin of the coordinate system being the origin of the position vectors). Certainly we could clarify in documentation what you can and can't do with the standard rotations, but that's also moving past documentation and into reference material (which is fine, but noting that it's a higher level of work).

The case where you have a local coordinate system is different though. E.g., if your measurements are oriented by a local North, East, Down system then at each point "North" will be a tangent to the sphere (or geoid, depending), as will "East". So the rotation won't match up with any of our defined systems, because the axes of the coordinate systems aren't aligned. 

I have an example of this that I can elaborate on later, for local magnetic field measured at Earth's surface in (North, East, Zenith) which is likely the same transformation needed here. I'lll give a concrete solution to _that_ problem later (when I'm at a computer and have a few extra minutes). The key is that you need to define the axes of your coordinate system (which are position dependent) in another, known system before you can start the transforms.

From the spacepy perspective I think this is something we could use in our library of examples/case studies. I think the general case may also be easier to integrate after the coordinates overhaul is finished and merged because it'll allow direct access to all the underlying rotation matrices.

**Comment by drsteve** [comment, marked as answer]
Okay, so the ultimate problem here is that (as I noted in my earlier comment) the coordinate system your data are in is probably a local (and hence position-dependent) system. There are a few paths to convert from one local system to another, but it's probably easiest to go through standard systems. Partly because this should be easier to follow, and partly because this is what I have demo code for already.

Part of the solution is implemented in [the gist here](https://gist.github.com/drsteve/626dc0473670d3a4755c626edcf7d6b7)

What the gist provides is a conversion between a local North-East-Down system and its parent (and vice versa). By that I mean, if local North and East are geographic then the parent coordinate system is geocentric geographic.

So to convert from local [geo north, geo east, nadir] to geocentric geographic the code defines the bases of the local coordinate system in the parent (i.e., z_hat = [0, 0, 1] in the local system, but is radial in the parent system). Then we can get rotation matrices to convert between local and parent systems.

To convert to a local system with different bases you should be able to convert to a different parent system (MAG), then back to a local system based on that parent. The direct route would be to figure out how to define the basis vectors of your target (local MAG) coordinate system in your local GEO system and generate that rotation instead. It would follow a similar prescription to the gist linked here, but it'd require details of the centered dipole axis. That's not directly available in code here until #296 is finished/merged.

**Reply by tiger2017** [reply]
Thank you! Yes, the vector data I had is in a local coordinate system, not North-East-Down system but East (for X)-North (for Y)-Up (for Z) system (North and East are tangent to the sphere surface and Up is normal to the sphere surface but outwards (like radial direction)) but I think they are pretty similar and can be just a positive-negative symbol for the Z axis component.

For the East-North-Up system converting to MAG coordinate, I guess the only thing I need to modify are just only two places:
In the function "def ned_to_parent":
enu = np.asarray(invec)[[1, 0, 2]] * np.array([1, 1, 1]), or just simply: enu = np.asarray(invec)

In the function "parent_to_ned":
ned = np.asarray(enu)[[1, 0, 2]] * np.array([1, 1, 1]), or just simply: ned = np.asarray(enu)

Am I right?

One following question is for the def "calc_transmat(lat, lon)", where the defined input parameters lat/lon here seems be different from the lat/lon input parameters in the other two functions I mentioned above, they seem not to be the location information of my studying data but should be something like rotation angles between GEO and MAG? Can you provide more details about the definition of the lat/lon in the function calc_transmat?

Also look forward to seeing a direct way about the rotation from GEO to MAG with the dependence of date-time after #296 is finished (I know there are some time-dependent dipole field model existing), which would be of a great help in solving the vector investigation problem. Thanks again.

**Reply by tiger2017** [reply]
Just realized that the rotation from GEO to MAG should not be time-dependent, but fixed during a non-long term (e.g. a few years probably).

**Reply by drsteve** [reply]
> Yes, the vector data I had is in a local coordinate system, not North-East-Down system but East (for X)-North (for Y)-Up (for Z) system (North and East are tangent to the sphere surface and Up is normal to the sphere surface but outwards (like radial direction)) but I think they are pretty similar and can be just a positive-negative symbol for the Z axis component.

> For the East-North-Up system converting to MAG coordinate, I guess the only thing I need to modify are just only two places:

The first line of `ned_to_parent` is converting North-East-Down to East-North-Up. The last last of `parent_to_ned` is doing the inverse. If you are starting with ENU then you should skip that step. Your suggestions of replacing the first line of `ned_to_parent` with `enu = np.asarray(invec)` will cover that. 

The rest of the steps are (using `ned_to_parent` as the example):
1. calculate transformation matrices based on the position of the input vector
2. apply the rotation (this is the dot product step)

> One following question is for the def "calc_transmat(lat, lon)", where the defined input parameters lat/lon here seems be different from the lat/lon input parameters in the other two functions I mentioned above, they seem not to be the location information of my studying data but should be something like rotation angles between GEO and MAG? Can you provide more details about the definition of the lat/lon in the function calc_transmat?

`calc_transmat` is calculating the transformation matrices between the position-dependent local system and the parent, the `lat` and `lon` used there are just the `lat` and `lon` values passed through from the calling `ned_to_parent` or `parent_to_ned` routine. These are just the latitude and longitude (in the parent system) of the origin for your local coordinate system as they're used to define the tangent plane and get the axes of the local system in the parent system.

> the rotation from GEO to MAG should not be time-dependent, but fixed during a non-long term (e.g. a few years probably).

That really depends on the accuracy you need for your problem. Over the course of a day, sure, the dipole axis can be considered fixed relative to the Earth's surface for most applications. But it depends on the problem. My default approach is to not make approximations that I don't need to, as it saves me the effort of justifying the assumptions and testing the accuracy loss due to them. So I would re-evaluate the dipole axis for each time and that way my workflow is the same regardless of whether I'm looking at an hour, a day, or a solar cycle. It's fairly common for folks to only update the dipole axis on a lower cadence (often daily, sometimes as low as 6 months).
